### PR TITLE
chop request functions into handy bits, reuse code

### PR DESCRIPTION
**Pull request by dothebart**
### Scope & Purpose

re-work V8 client functions to share code and behave more similar

- [x] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] Backports required for: 3.7, 3.8

### CODE REVIEW

File: arangosh/Shell/V8ClientConnection.cpp [hunk 1]
```diff
@@ -1735,52 +1735,98 @@ v8::Local<v8::Value> V8ClientConnection::patchData(
...
-        req->header.contentType(fu::ContentType::Json);
+        request.header.contentType(fu::ContentType::Json);
       } else {
-        req->header.contentType(fu::ContentType::Custom);
+        request.header.contentType(fu::ContentType::Custom);
       }
     } else if (boost::iequals(StaticStrings::Accept, pair.first)) {
-      req->header.acceptType(fu::ContentType::Custom);
+      request.header.acceptType(fu::ContentType::Custom);
     }
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1787]: I am still not convinced that we need all this here, given that fuerte's header.addMeta(...) should do it already.

File: arangosh/Shell/V8ClientConnection.cpp [hunk 2]
```diff
@@ -1790,41 +1836,135 @@ v8::Local<v8::Value> V8ClientConnection::requestData(
...
-    } else {
-      req->header.acceptType(fu::ContentType::VPack);
+  return true;
+}
+
+bool canParseResponse(fu::Response const& response) {
+  return (response.contentType() != fuerte::ContentType::Custom) &&
+    (response.isContentTypeVPack() || response.isContentTypeJSON()) &&
+    (response.contentEncoding() == fuerte::ContentEncoding::Identity) &&
+    (response.payload().size() > 0);
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1868]:
```suggestion
  return (response.isContentTypeVPack() || response.isContentTypeJSON()) &&
         response.contentEncoding() == fuerte::ContentEncoding::Identity &&
         response.payload().size() > 0;
```

File: arangosh/Shell/V8ClientConnection.cpp [hunk 3]
```diff
@@ -1790,41 +1836,135 @@ v8::Local<v8::Value> V8ClientConnection::requestData(
...
+                v8::Boolean::New(isolate, true)).FromMaybe(isError);
+    result->Set(context,
+                TRI_V8_STD_STRING(isolate, StaticStrings::ErrorNum),
+                v8::Integer::New(isolate, lastHttpReturnCode)).FromMaybe(false);
+    result->Set(context,
+                TRI_V8_STD_STRING(isolate, StaticStrings::ErrorMessage),
+                TRI_V8_STD_STRING(isolate, message)).FromMaybe(false);
+}
+
+void setResultMessage(v8::Isolate* isolate,
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1927]: It seems there are 2 `setResultMessage` functions now, one with a message parameter and one without. Can we move them close together in code so this becomes more obvious to the code reader?

Comment (dothebart) [reply arangosh/Shell/V8ClientConnection.cpp:1926]: the one even calls the other?

File: arangosh/Shell/V8ClientConnection.cpp (hunk 1 above)
Comment (dothebart) [reply arangosh/Shell/V8ClientConnection.cpp]: its unneccessary, completely removed.
flipped sequence for `forceJson` so it is now an overrideable default.

File: tests/js/client/shell/shell-foxx-interface.js [hunk 4]
```diff
@@ -168,8 +170,9 @@ function foxxInterfaceSuite () {
     testFoxxInterfaceHeadBinary: function () {
       let res = arango.HEAD_RAW(binUrl);
       assertEqual(res.code, 200, res.parsedBody);
-      assertTrue(res.body instanceof Buffer);
-      assertEqual(res.body.length, 0);
+      // we don't create bodies for head requests anymore:
+      //assertTrue(res.body instanceof Buffer);
+      //assertEqual(res.body.length, 0);
```
Comment (pluma) [inline tests/js/client/shell/shell-foxx-interface.js:175]: This should probably be `assertUndefined(res.body);` like in the other tests then.

**Review by pluma — CHANGES_REQUESTED** [review]

File: arangosh/Shell/V8ClientConnection.cpp [hunk 5]
```diff
@@ -1893,10 +1879,14 @@ v8::Local<v8::Value> parseReplyBodyToV8(fu::Response const& response,
 v8::Local<v8::Value> translateResultBodyToV8(fu::Response const& response,
                                              v8::Isolate* isolate) {
   auto responseBody = response.payload();
-
-  if ((response.contentEncoding() == fuerte::ContentEncoding::Identity) ||
-      response.isContentTypeHtml() ||
-      response.isContentTypeText()) {
+  if ((responseBody.size() > 0) &&
```
Comment (jsteemann) [inline arangosh/Shell/V8ClientConnection.cpp:1882]: This change means that for any response body of length 0, we will create a Buffer as the result and no empty string. Is that intentional?

File: tests/js/client/shell/shell-foxx-interface.js [hunk 6]
```diff
@@ -168,9 +170,11 @@ function foxxInterfaceSuite () {
     testFoxxInterfaceHeadBinary: function () {
       let res = arango.HEAD_RAW(binUrl);
       assertEqual(res.code, 200, res.parsedBody);
-      assertTrue(res.body instanceof Buffer);
-      assertEqual(res.body.length, 0);
-      
+      // we don't create bodies for head requests anymore:
+      //assertTrue(res.body instanceof Buffer);
+      //assertEqual(res.body.length, 0);
```
Comment (jsteemann) [inline tests/js/client/shell/shell-foxx-interface.js:175]: If these assertions are not executed, let's remove them.

File: arangosh/Shell/V8ClientConnection.cpp (hunk 5 above)
Comment (dothebart) [reply arangosh/Shell/V8ClientConnection.cpp:1882]: yes, this was the original behaviour, tests asserted this.

Comment (dothebart) [reply arangosh/Shell/V8ClientConnection.cpp]: removed.

File: CHANGELOG [hunk 7]
```diff
@@ -1,6 +1,12 @@
 devel
 -----
 
+* change arangosh client behaviour:
+ - *_RAW methods will never add a `body` to HEAD requests
```
Comment (jsteemann) [inline CHANGELOG:5]:
```suggestion
 - *_RAW methods will never add a `body` to HEAD responses
```

File: CHANGELOG [hunk 8]
```diff
@@ -1,6 +1,12 @@
 devel
 -----
 
+* change arangosh client behaviour:
```
Comment (jsteemann) [inline CHANGELOG:4]:
```suggestion
* change arangosh client behavior:
```

File: CHANGELOG [hunk 9]
```diff
@@ -1,6 +1,12 @@
...
+ - *_RAW methods will never add a `body` to HEAD requests
+ - *_RAW methods will now allways make velocypack replies SlowBuffers
```
Comment (jsteemann) [inline CHANGELOG:6]:
```suggestion
 - *_RAW methods will now always return velocypack-typed responses in Buffers
```

**Review by jsteemann — APPROVED** [review]
LGTM

**Review by pluma — APPROVED** [review]
JS LGTM

**Pull request merged**